### PR TITLE
Challenge caching

### PR DESCRIPTION
**Pull request by Anze1m**
# Checklist:

- [x] Code follows the style guidelines
- [x] Documentation updated
- [x] Corner cases evaluated and handled
- [x] Automatic tests added/update
- [x] Automatic tests pass
- [x] Software runs (locally or hosted)
- [x] Sanity check passed
- [x] Self-review passed

### CODE REVIEW

**Comment by Anze1m** [comment]
Instead of caching whole interval at once, cache events one by one. Sort by `block`, sort by `logIndex`, decide what event is that and apply to cached state.

File: test/services/challenges_repository.js [hunk 1]
```diff
@@ -183,38 +140,55 @@ describe('Challenges repository', () => {
...
-        [{challengeId: 'resolved', blockNumber: 3, resolverId: 'someResolverId'}],
-        [{challengeId: 'timedOut', blockNumber: 5}],
-        challengeWrapperMock.defaultAddress
+      expect(challengeWrapperMock.challenges).to.be.calledWith(latestBlock + 1, latestBlock + 3);
+      expect(challengeWrapperMock.resolvedChallenges).to.be.calledWith(latestBlock + 1, latestBlock + 3);
+      expect(challengeWrapperMock.timedOutChallenges).to.be.calledWith(latestBlock + 1, latestBlock + 3);
+      expect(challengesRepository.lastSavedBlock).to.equal(latestBlock + 3);
+    });
+
+    it('adds new challenges to cache, decreases active count on resolved and removes timeouted', async () => {
```
Comment (krzysztof-jelski) [inline test/services/challenges_repository.js]: timeouted -> timed out

File: test/services/active_challenges_cache.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      {blockNumber: 5, logIndex: 7}
+    ];
+
+    it('in increasing blockNumber and logIndex order', () => {
+      expect(activeChallengesCache.sortChronologically(events)).to.deep.equal([{blockNumber: 5, logIndex: 2}, {blockNumber: 5, logIndex: 7}, {blockNumber: 9, logIndex: 4}]);
+    });
+  });
+
+  describe('Applies multiple changes', () => {
+    it('Applies in correct order', () => {
```
Comment (krzysztof-jelski) [inline test/services/active_challenges_cache.js]: This test is very hard on the reader. You can't quickly figure out why the expected results are what they are. Consider splitting into a couple of smaller tests.

File: test/services/challenges_repository.js [hunk 3]
```diff
@@ -183,38 +140,55 @@ describe('Challenges repository', () => {
       configWrapperMock = {
         challengeDuration: sinon.stub().resolves(challengeDuration)
       };
-      challengesRepository = new ChallengesRepository(challengeWrapperMock, configWrapperMock);
-      sinon.spy(challengesRepository, 'extractChallengeFromEvent');
-      sinon.stub(challengesRepository, 'filterOutFinishedChallenges').returns(notFinishedEvents);
-      sinon.stub(challengesRepository, 'sortChallenges').returns(sortedEvents);
+      activeChallengesCacheMock = {
```
Comment (krzysztof-jelski) [inline test/services/challenges_repository.js]: I would prefer using the real thing in this test (`new ActiveChallengesCache()`).

File: src/services/active_challenges_cache.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    if (this.has(challenge.challengeId)) {
+      this.activeChallengesDict[challenge.challengeId].count --;
+      if (this.get(challenge.challengeId).count <= 0) {
+        delete this.activeChallengesDict[challenge.challengeId];
+      }
+    }
+  }
+
+  applyIncomingChallengeEvents(startedChallenges, resolvedChallenges, timedOutChallenges) {
+    const startedChallengesActions = startedChallenges.map((challengeAction) => this.addType(challengeAction, 'issued'));
```
Comment (krzysztof-jelski) [inline src/services/active_challenges_cache.js]: Store actions instead of strings, it will simplify code.

File: test/services/active_challenges_cache.js (hunk 2 above)
Comment (Anze1m) [reply test/services/active_challenges_cache.js]: To be done.

File: test/services/challenges_repository.js (hunk 3 above)
Comment (Anze1m) [reply test/services/challenges_repository.js]: Why? `ActiveChallengesCache ` is tested separately.

File: test/services/active_challenges_cache.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+  beforeEach(() => {
+    activeChallengesCache = new ActiveChallengesCache();
+  });
+
+  function aChallenge() {
+    return exampleChallenge;
+  }
+
+  function aChallengeWith(params) {
```
Comment (spherefoundry) [inline test/services/active_challenges_cache.js]: you can merge `aChallenge` and `aChallengeWith` methods by using a params parameters with a default of {};

File: test/services/active_challenges_cache.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  let activeChallengesCache;
+  const exampleChallenge = {
+    challengeId: '0xc0ffee',
+    sheltererId: '0xbeef',
+    bundleId: '0xdeadbeef',
+    count: 12,
+    blockNumber: 1,
+    logIndex: 0
+  };
+  const anotherChallengeId = {challengeId: 'anotherChallengeId'};
```
Comment (spherefoundry) [inline test/services/active_challenges_cache.js]: The name is somewhat confusing. I would expect a hex string, but it's a dict 🤔

File: test/services/active_challenges_cache.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    activeChallengesCache.add(aChallengeWith({challengeId: '0xbeef', blockNumber: 3}));
+    activeChallengesCache.add(aChallengeWith({challengeId: '0xdeadface', blockNumber: 2, logIndex: 1}));
+    activeChallengesCache.add(aChallengeWith({challengeId: '0xcafebabe', blockNumber: 2, logIndex: 0}));
+
+    expect(activeChallengesCache.activeChallenges).to.deep.equal([
+      aChallengeWith({challengeId: '0xc0ffee', blockNumber: 1}),
+      aChallengeWith({challengeId: '0xcafebabe', blockNumber: 2, logIndex: 0}),
+      aChallengeWith({challengeId: '0xdeadface', blockNumber: 2, logIndex: 1}),
+      aChallengeWith({challengeId: '0xbeef', blockNumber: 3}),
+      aChallengeWith({challengeId: '0xbaddad', blockNumber: 4})
```
Comment (spherefoundry) [inline test/services/active_challenges_cache.js]: IMO: there is a lot of repetition here with all the `aChallenge...` calls. Maybe assign them to local variables? This will make the method a little bit longer, but IMO more readable.

File: src/services/challenges_repository.js [hunk 8]
```diff
@@ -46,14 +47,17 @@ export default class ChallengesRepository {
     this.activeChallengesCache.applyIncomingChallengeEvents(startedChallenges, resolvedChallenges, timedOutChallenges);
   }
 
-  async updateBlockInfo() {
+  async getBlockInfo() {
     const challengeDuration = await this.configWrapper.challengeDuration();
```
Comment (krzysztof-jelski) [inline src/services/challenges_repository.js]: We can save one additional call to parity if we did this only in case when there is no lastSavedBlock (rhs of below conditional)
